components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/LegacyJobOrchestrator.cs

```diff
@@ -249,113 +246,134 @@ public class LegacyJobOrchestrator : IJobOrchestrator, IPublisherConfigServices,
             return BitConverter.ToString(checksum).Replace("-", string.Empty);
         }
 
+        /// <summary>
+        /// Deserialize string representing published nodes file into PublishedNodesEntryModel entries and
+        /// run schema validation if that is enables.
+        /// </summary>
+        /// <param name="content"></param>
+        /// <returns></returns>
+        private IEnumerable<PublishedNodesEntryModel> DeserializePublishedNodes(string content) {
+            if (!File.Exists(_legacyCliModel.PublishedNodesSchemaFile)) {
+                _logger.Information("Validation schema file {PublishedNodesSchemaFile} does not exist or is disabled, ignoring validation of {publishedNodesFile} file...",
+                _legacyCliModel.PublishedNodesSchemaFile, _legacyCliModel.PublishedNodesFile);
+                return _publishedNodesJobConverter.Read(content, null);
+            }
+            else {
+                using (var fileSchemaReader = new StreamReader(_legacyCliModel.PublishedNodesSchemaFile)) {
+                    return _publishedNodesJobConverter.Read(content, fileSchemaReader);
+                }
+            }
+        }
+
+        private bool IsSameDataSet(PublishedNodesEntryModel entry1, PublishedNodesEntryModel entry2) {
+            return entry1.HasSameGroup(entry2)
+                && entry1.DataSetWriterId == entry2.DataSetWriterId
+                && entry1.DataSetPublishingInterval == entry2.DataSetPublishingInterval;
+        }
+
+        private void RefreshJobs(IEnumerable<PublishedNodesEntryModel> entries) {
+            var availableJobs = new Dictionary<string, JobProcessingInstructionModel>();
+            var assignedJobs = new Dictionary<string, JobProcessingInstructionModel>();
+
+            _lockJobs.Wait();
```

Comment:
>_lockJobs.Wait(); [](http://example.com/codeflow?start=12&length=17)

let's lock the jobs array after we do the conversion since this is quite time consuming and is independent of the jobs assignment

Reply:
First operation in `try` block reads values from `_assignedJobs`, so that should be under the lock. And releasing the lock in between those two operations may result in a change to `_assignedJobs` that will not be reflected in the resulting refreshed jobs, so I would like to keep the locking mechanism as it is here unless there are some other concerns.

Reply:
yeah, I meant we can move 
```var jobs = _publishedNodesJobConverter.ToWriterGroupJobs(entries, _legacyCliModel);```
 in the beginning - outside the lock, since it has no order dependency on 
```
 var jobIdDict = _assignedJobs
                    .Select(kvp => Tuple.Create(kvp.Value.Job.Id, kvp.Key))
                    .ToDictionary(tuple => tuple.Item1);
```

Reply:
Ah, sure. I will move job conversion out of the lock block.

Reply:
Done.